File: code/chapter5/xirr.R
NPV = function(cashflow_dates, cashflow_amounts, discount_rate){
  stopifnot(length(cashflow_dates)==length(cashflow_amounts))
  days_in_year = 365 # we assume there are 365 days in each year
  times= as.numeric(difftime(cashflow_dates, cashflow_dates[1], units="days"))/days_in_year
  sum(cashflow_amounts/(1+discount_rate)^times)
}

xirr = function(cashflow_dates, cashflow_amounts){
  stopifnot(length(cashflow_dates)==length(cashflow_amounts))
  # we use uniroot function in R for root-finding
  uniroot(f=function(x) NPV(cashflow_dates,cashflow_amounts,x), interval= c(-1,100))$root
}

File: code/chapter5/xirr.py
from datetime import datetime
from scipy.optimize import root

def NPV(cashflow_dates, cashflow_amounts, discount_rate):
  assert len(cashflow_dates)==len(cashflow_amounts), "inconsistent lengths of cashflows dates and amounts"
  days_in_year=365.0
  times = [(e-cashflow_dates[0]).days/days_in_year for e in cashflow_dates]
  return sum(cashflow_amounts[i]/(1.0+discount_rate)**e for i,e in enumerate(times))

def xirr(cashflow_dates, cashflow_amounts):
  # we use the scipy.optimize.root function to find the root
  return root(fun=lambda x:NPV(cashflow_dates,cashflow_amounts,x),x0=[0.0]).x
